src/helpers/cloneChildren.js

```diff
@@ -1,38 +1,113 @@
 /*eslint func-style:0*/
 import React from 'react';
+import identity from './identity';
+import compose from './compose';
+import warning from 'warning';
 
 /**
- * Clones a child subtree, when we encounter a component that passes our
- * predicate pass it down additional props.
+ * Rule for cloning something at leaf level like some text
+ */
+const leafRule = {
+    predicate: (child) => typeof child !== 'object' || child === null,
+    clone: identity
+}
+
+/**
+ * Allows default recursion into an element that has children.
  *
- * @param  {Function} predicate A predicate function which recives the child
- * @param  {Function} getProps  A function which recives the component and
- * returns an object which gets merged into the props of the component
- * @param  {Function} children The children to iterate over
- * @return {Object} The cloned children
+ * @param {array} rules on how to clone individual elements
+ * @returns {Object} rule for cloning recursively
  */
-export default function cloneChildren(predicate, getProps, children) {
-    if (typeof children !== 'object' || children === null) {
-        return children;
+function createRecursiveRule(rules) {
+    return {
+        predicate: () => true,
+        clone: (child, childNames) => {
+            return React.cloneElement(child, {},
+                cloneChildren(rules, child.props && child.props.children, childNames));
+        }
     }
+}
 
-    return React.Children.map(children, (child) => {
-        if (typeof child !== 'object' || child === null) {
-            return child;
+/**
+ * A common function for cloning Errors element that takes care of injecting
+ * required error data
+ *
+ * @param {Object} errors aaaaa
+ * @param {Object} fieldErrors bbbb
+ * @return {Object} rule for cloning Errors element
+ */
+export function createErrorsRule({ errors = [], fieldErrors = {} }) {
```

Comment:
It would be nice to keep our signatures free of destructuring to be consistent. `createErrorsRule` and `createFormableRule` both destructure, however `getFormableComponentProperties` does not and has pretty much the same signature. 

I think it would be more explicit to remove the destructuring and make the caller explicitly pass in what is needed. For example, in `Fieldset` and `Form`, we wouldn't pass `this.props` / `this.state`, we would pass in the errors we get form our parents. 

Just a thought.